### ISSUE TITLE
KubeSpawner did not attempt to delete a pending pod?

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/kubespawner#240]
## TL;DR
A `.stop()` request to KubeSpawner (which arise from a DELETE request to `/hub/api/users/erik/server` I presume) should result in deleting a pending user pod. It does not make much sense to me that it currently instead comes up with this error response:
```
{"status": 400, "message": "erik is not running (pending: spawn)"}
```

## About
After acessing the `/hub/api` endpoint using webrequests, and accessing the hub's API [as defined here on Swagger](http://petstore.swagger.io/?url=https://raw.githubusercontent.com/jupyterhub/jupyterhub/master/docs/rest-api.yml), I realized that a user that stuck pending spawn, can't be deleted by requests to the hub's API.

Shouldn't a DELETE request (or "stop server please") actually delete a pending pod?

I realize this is the reason I've experienced an inability to stop pending pods from the hub admin panel, which isn't a great experience.

## My terminal
```sh
# POST request to <IP>/hub/api/users/erik
{"kind": "user", "name": "erik", "admin": false, "groups": [], "server": null, "pending": null, "created": "2018-08-11T01:38:27.345309Z", "last_activity": null, "servers": null}

# POST request to <IP>/hub/api/users/erik/server
# ...

# kubectl get pod jupyter-erik
NAME                              READY     STATUS    RESTARTS   AGE
jupyter-erik                      0/1       Pending   0          3m

# DELETE request to <IP>/hub/api/users/erik/server
{"status": 400, "message": "erik is not running (pending: spawn)"}

# kubectl get pod jupyter-erik
NAME                              READY     STATUS    RESTARTS   AGE
jupyter-erik                      0/1       Pending   0          3m

# kubectl get pod jupyter-erik --- when KubeSpawner has given up
Error from server (NotFound): pods "jupyter-erik" not found
```

---

## Hub's logs
```
# 1. Creating a user
# -------------------------------------
[W 2018-08-11 01:40:43.028 JupyterHub base:679] User erik is slow to start (timeout=10)
[I 2018-08-11 01:40:43.030 JupyterHub log:158] 202 POST /hub/api/users/erik/server (tests@10.20.0.1) 10114.30ms
[I 2018-08-11 01:40:53.066 JupyterHub proxy:301] Checking routes
[I 2018-08-11 01:41:53.065 JupyterHub proxy:301] Checking routes
[I 2018-08-11 01:42:53.070 JupyterHub proxy:301] Checking routes
[W 180811 01:42:53 cull_idle_servers:128] Not culling server erik with pending spawn
[I 2018-08-11 01:42:53.357 JupyterHub log:158] 200 GET /hub/api/users (cull-idle@127.0.0.1) 37.27ms

# 2. Deleting the user, but fails
# -------------------------------------
[W 2018-08-11 01:43:11.220 JupyterHub web:1667] 400 DELETE /hub/api/users/erik/server (10.20.0.1): erik is not running (pending: spawn)
[W 2018-08-11 01:43:11.221 JupyterHub log:158] 400 DELETE /hub/api/users/erik/server (tests@10.20.0.1) 19.52ms
[I 2018-08-11 01:43:53.060 JupyterHub proxy:301] Checking routes
[I 2018-08-11 01:44:53.059 JupyterHub proxy:301] Checking routes

# 3. KubeSpawner gave up and deleted the pod
# -------------------------------------
[W 2018-08-11 01:45:33.077 JupyterHub user:468] erik's server failed to start in 300 seconds, giving up
[I 2018-08-11 01:45:33.080 JupyterHub spawner:1612] Deleting pod jupyter-erik
[E 2018-08-11 01:45:33.165 JupyterHub gen:974] Exception in Future <Task finished coro=<BaseHandler.spawn_single_user.<locals>.finish_user_spawn() done, defined at /usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py:619> exception=TimeoutError('Timeout',)> after timeout
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/tornado/gen.py", line 970, in error_callback
        future.result()
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py", line 626, in finish_user_spawn
        await spawn_future
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/user.py", line 486, in spawn
        raise e
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/user.py", line 406, in spawn
        url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
    tornado.util.TimeoutError: Timeout
    
[I 2018-08-11 01:45:53.062 JupyterHub proxy:301] Checking routes

# 4. The pod reflector got sad and restarted, thinking it missed a startup
[E 2018-08-11 01:45:54.108 JupyterHub spawner:1580] Pod jupyter-erik never showed up in reflector, restarting pod reflector
[I 2018-08-11 01:45:54.138 JupyterHub reflector:187] watching for pods with label selector='component=singleuser-server' in namespace jh
```

**Comment by minrk** [comment]
> Shouldn't a DELETE request (or "stop server please") actually delete a pending pod?

This is entirely a JupyterHub issue. JupyterHub refuses to delete Spawners that are in the middle of starting as a protection for Spawner authors, so that they don't need to be able to deal with `.stop()` being called at every possible arbitrary point in the middle of an async `.start()`. JupyterHub, for safety, waits for `.start()` to either succeed or fail before allowing `.stop()` to be called.

**Comment by consideRatio** [comment]
Ah, perhaps not much to do about this then! Closing.
